Binder/AIDL content updates

randomPoison opened the pull request:
So far this PR makes two broad improvements to the Android course materials:

* It fleshes out the existing course materials by adding more speaker notes and showing more of the generated code.
* It adds a new section going over how AIDL types are supported by Rust and mapped to Rust types.

There are still additional changes to do (add an exercise, add a section on async binder), but those can be split into separate PRs if necessary.

This PR is still WIP because I need to test the code examples around the Binder types; Some of them are basically just psuedo-code at the moment.

rinon commented on src/android/aidl/birthday-service.md:
```diff
@@ -0,0 +1,3 @@
+# Birthday Service Tutorial
+
+We doing an example! We're going to create a Birthday Service that does... something.
```
Assuming this is going to be improved? Title slides should serve some purpose (or we can skip it).

randomPoison replied:
Yup! Wording has been improved. It's still a quick slide, but now it at least introduces the section.

rinon commented on src/android/aidl/birthday-service/birthday_service/Android.bp:
Do we need to move the sources? If there's a concrete reason to, then cool, but the double directory is a bit confusing.

rinon commented on src/android/aidl/birthday-service/birthday_service/aidl/com/example/birthdayservice/IBirthdayService.aidl:
```diff
@@ -0,0 +1,35 @@
+    /** Generate a Happy Birthday message. */
+    String wishHappyBirthday(String name, int years);
+
+    /** The same thing, but using more Binder features. */
```
Probably worth being specific about which feature, in this case, callbacks?

randomPoison replied:
Yup, I've changed the wording here to be clearer, and the example has been simplified to only use the one binder feature (sending a binder object).

rinon commented on src/android/aidl/birthday-service/client.md:
```diff
@@ -0,0 +1,44 @@
+  that the client has connected to.
+    * `Strong` is a custom smart pointer type for Binder. It handles both an
+      in-process ref count for the service trait object, and the global Binder
+      ref count that tracks how many clients are connected to a given service.
```
It's more "have a reference to the Binder object," because objects != services.

randomPoison replied:
I've tweaked the wording. Let me know if it's still not quite right.

rinon reviewed (commented):
This is looking really great! More depth than was there before, and I like the new notes you've incorporated. I think that covers most of what I've added off slides when I've taught this, but I'll think if there's anything else we should add.

ahomescu commented on src/android/aidl/birthday-service/birthday_service/aidl/com/example/birthdayservice/IBirthdayInfoProvider.aidl:
```diff
@@ -1,4 +1,4 @@
-// Copyright 2022 Google LLC
+// Copyright 2024 Google LLC
```
Nit: `2022-2024` or just keep the original year?

randomPoison replied:
This is a new file I created, so I thought we'd want to use 2024 in the comment since that's when it's written?

mgeisler replied:
I think using 2024 is correct for new files. My understanding is that we leave the years alone in existing files.

Way back when I first read about licenses for open source projects, the advice was to update the year of files that have received "significant changes" or so. Many projects just updated the year of every file in the beginning of the year. It's now my understanding that we don't do that, we just leave the files alone and only add new notices to new files.

ahomescu commented on src/android/aidl/birthday-service/birthday_service/src/client.rs:
```diff
@@ -14,8 +14,11 @@
+use com_example_birthdayservice::aidl::com::example::birthdayservice::BirthdayInfo::BirthdayInfo;
+use com_example_birthdayservice::aidl::com::example::birthdayservice::IBirthdayInfoProvider::IBirthdayInfoProvider;
 use com_example_birthdayservice::aidl::com::example::birthdayservice::IBirthdayService::IBirthdayService;
 use com_example_birthdayservice::binder;
```
```suggestion
use com_example_birthdayservice::binder::{self, Strong};
```

ahomescu commented on src/android/aidl/birthday-service/birthday_service/src/client.rs:
```diff
@@ -14,8 +14,11 @@
+use com_example_birthdayservice::aidl::com::example::birthdayservice::IBirthdayInfoProvider::IBirthdayInfoProvider;
 use com_example_birthdayservice::aidl::com::example::birthdayservice::IBirthdayService::IBirthdayService;
 use com_example_birthdayservice::binder;
+use binder::Strong;
```
Does this import from the `binder` crate, or the module on the previous line? Either way, we should probably avoid the ambiguity in case `rustfmt` reorders these lines.

randomPoison replied:
I unified it to the binder module the way you suggested.

ahomescu commented on src/android/aidl/birthday-service/birthday_service/src/client.rs:
```diff
@@ -36,7 +39,47 @@ fn main() -> Result<(), binder::Status> {
+    // ANCHOR: wish_with_provider
+    // Perform the same operation by sending an `InfoProvider` object as a `dyn
+    // IBirthdayInfoProvider`.
+    let provider = Strong::new(Box::new(InfoProvider { name, age: years as u8 }));
```
```suggestion
    let provider = BnBirthdayInfoProvider::new_binder(InfoProvider { name, age: years as u8 }, BinderFeatures::default());
```
or something like that.

ahomescu commented on src/android/aidl/birthday-service/birthday_service/src/client.rs:
```diff
@@ -36,7 +39,47 @@ fn main() -> Result<(), binder::Status> {
+    service.wishWithProvider(&provider);
+    // ANCHOR_END: wish_with_provider
+
+    // TODO: Perform the same operation but passing the provider as an `SpIBinder`.
```
This one should be simple: `provider.as_binder()`.

ahomescu commented on src/android/aidl/birthday-service/birthday_service/src/lib.rs:
```diff
@@ -14,6 +14,8 @@
+use binder::{ParcelFileDescriptor, SpIBinder, Strong};
+use com_example_birthdayservice::aidl::com::example::birthdayservice::IBirthdayInfoProvider::IBirthdayInfoProvider;
 use com_example_birthdayservice::aidl::com::example::birthdayservice::IBirthdayService::IBirthdayService;
 use com_example_birthdayservice::binder;
```
See comment from the other file: we probably want to import the `binder::` types from here instead.

randomPoison replied:
Done!

ahomescu commented on src/android/aidl/birthday-service/service-bindings.md:
```diff
@@ -0,0 +1,30 @@
+# Generated Service API
+
+Binder generates a trait corresponding to the interface defintion. trait to talk
```
Typo: defintion.

ahomescu commented on src/android/aidl/types.md:
```diff
@@ -0,0 +1,8 @@
+
+* Primitive types map (mostly) to idiomatic Rust types.
+* Collection types like slices, `Vec`s and string types are supported.
+* References to AIDL objects can be sent between file handles.
```
The phrasing here is a little clunky. Do you mean you can send objects, or file handles?

randomPoison replied:
I think my brain farted while typing that sentence. Fixed!

ahomescu commented on src/android/aidl/types/file-descriptor.md:
````diff
@@ -0,0 +1,17 @@
+
+```rust,ignore
+let info_file = File::open("info.txt")?;
+let msg = service.wishHappyBirthday(info_file)?; // TODO: Is this correct? How do we convert to a `ParcelFileDescriptor`?
````
```suggestion
let pfd = ParcelFileDescriptor::new(info_file);
let msg = service.wishHappyBirthday(pfd)?;
```

ahomescu commented on src/android/aidl/types/parcelables.md:
```diff
@@ -0,0 +1,33 @@
+
+* Parcelables work similarly to serde types in the broader Rust ecosystem: You
+  get a concrete Rust type to work with in your Rust code, and that type can be
+  serializied without you needing to write any serialization logic yourself.
```
Typo: serializied.

ahomescu commented on src/android/aidl/types/parcelables.md:
````diff
@@ -0,0 +1,33 @@
+**client/src/main.rs**:
+
+```rust,ignore
+let msg = service.wishHappyBirthday(BirthdayInfo {
````
I don't remember off the top of my head: is the argument by value or by reference?

randomPoison replied:
By reference, since it's an `in` parameter. The code example has been updated.

randomPoison commented:
@ahomescu @rinon I've finished updating the example code and have addressed the feedback from your first review. Give this another look over when you have a chance.

ahomescu commented on src/android/aidl/birthday_service/src/client.rs:
```diff
@@ -12,19 +12,19 @@
-}
-
 /// Call the birthday service.
 fn main() -> Result<(), binder::Status> {
```
You're unwrapping the I/O errors but not the Binder ones? I'm thinking you could just remove the `Result` here and use `unwrap` everywhere.

An alternative that would also make things consistent but with more effort: declare a new error type that covers all possible kinds, e.g., using the `thiserror` crate.

randomPoison replied:
Actually I can use `Box<dyn Error>`, which allows us to unify all the error types without using `thiserror` or `anyhow` (which is the crate that I'd otherwise use for a binary like this).

ahomescu commented on src/android/aidl/birthday_service/src/client.rs:
```diff
@@ -34,8 +34,65 @@ fn main() -> Result<(), binder::Status> {
+    writeln!(file, "{years}").unwrap();
+
+    // Create a `ParcelFileDescriptor` from the file and send it.
+    let file = ParcelFileDescriptor::new(file);
```
Do you need `file.into()` here? We switched PFD to take an `OwnedFd` recently.

randomPoison replied:
Doesn't seem like it? I'm able to `repo sync` and build against latest without the `.into()`. Would those changes be on a particular AOSP branch? I'm on whatever main branch gets checked out by default.

ahomescu replied:
If you're on `main`  and  you tested the current code there, that's good enough for me.

ahomescu commented on src/android/aidl/birthday_service/src/lib.rs:
```diff
@@ -26,4 +30,61 @@ impl IBirthdayService for BirthdayService {
+        &self,
+        provider: &SpIBinder,
+    ) -> binder::Result<String> {
+        use binder::binder_impl::Proxy;
```
Isn't this re-exported by `binder`? If it's not but it's needed, we should fix that upstream.

randomPoison replied:
It does not currently appear to be re-exported.

ahomescu replied:
Never mind, see the next comment.

ahomescu commented on src/android/aidl/birthday_service/src/lib.rs:
```diff
@@ -26,4 +30,61 @@ impl IBirthdayService for BirthdayService {
+    ) -> binder::Result<String> {
+        let mut info_file = info_file
+            .as_ref()
+            .try_clone()
```
Might be useful or interesting to explain what's going on here (either verbally, in the notes, or in a code comment here): `ParcelFileDescriptor` wraps a `std::os::fd::OwnedFd` but we want a `File`. We need to convert one into the other.

randomPoison replied:
This code isn't included in the slides, but I'll add a comment explaining the conversion.

ahomescu commented on src/android/aidl/example-service/interface.md:
````diff
@@ -0,0 +1,28 @@
+{{#include ../birthday_service/aidl/Android.bp}}
+```
+
+Add `vendor_available: true` if your AIDL file is used by a binary in the vendor
````
Or `vendor: true` is it's _only_ for vendor, but the audience might already know that.

randomPoison replied:
I'm not sure we even need this comment about `vendor_available` in this slide to begin with since it's more of a general AOSP thing and isn't really relevant to anything else we talk about. I'm removing it.

ahomescu commented on src/android/aidl/example-service/server.md:
```diff
@@ -0,0 +1,36 @@
+
+1. Create an instance of your service type (`BirthdayService`).
+1. Wrap the service object in corresponding `Bn*` type (`BnBirthdayService` in
+   this case). This type is generated by Binder and provides the common binder
```
Minor spelling nit: "Binder" and "binder" in the same sentence? Was that intentional?

randomPoison replied:
Nah, that was an accident. I've normalized it to "Binder" everywhere.

ahomescu commented on src/android/aidl/types/arrays.md:
```diff
@@ -0,0 +1,19 @@
+| ---------------------- | ------------- |
+| `in` argument          | `&[T]`        |
+| `out`/`inout` argument | `&mut Vec<T>` |
+| Return                 | `Vec<T>`      |
```
Also parcelable fields.

randomPoison replied:
What about parcelable fields? I'm not sure how that's relevant on this slide.

ahomescu replied:
I'm saying if you have a parcelable field that's an array, it will also use a `Vec<T>`. E.g.
```
parcelable Foo {
    int a[];
}
```

randomPoison replied:
Ah, that makes sense. I think I'll keep the slide specific to arrays in function signatures, but I'll add this to the speaker notes.

ahomescu commented on src/android/aidl/types/file-descriptor.md:
```diff
@@ -0,0 +1,38 @@
+# Sending Files
+
+Files can be sent between Binder clients/servers using the
```
Not sure if it should go here or in the notes, but it might be worth mentioning: other kinds of FDs can be wrapped, like TCP, UDP and UNIX sockets (e.g. https://doc.rust-lang.org/stable/std/os/fd/struct.OwnedFd.html#impl-From%3CUnixStream%3E-for-OwnedFd). Of particular interest on Android: shared memory buffers called DMA-BUFs.

randomPoison replied:
Good idea, comment added.

ahomescu commented on src/android/aidl/types/parcelables.md:
```diff
@@ -0,0 +1,38 @@
+
+- Parcelables work similarly to serde types in the broader Rust ecosystem: You
+  get a concrete Rust type to work with in your Rust code, and that type can be
+  serializied without you needing to write any serialization logic yourself.
```
```suggestion
  serialized without you needing to write any serialization logic yourself.
```

ahomescu commented on src/android/aidl/types/parcelables.md:
```diff
@@ -0,0 +1,38 @@
+
+<details>
+
+- Parcelables work similarly to serde types in the broader Rust ecosystem: You
```
IMHO they're mostly like C (or Rust) structs.

randomPoison replied:
Good point. I think the comment here is more confusing than helpful so I've removed it.

ahomescu commented on src/android/aidl/types/primitives.md:
```diff
@@ -0,0 +1,14 @@
+| --------- | --------- | ----------------------------------- |
+| `boolean` | `bool`    |                                     |
+| `byte`    | `i8`      | Note that bytes are signed.         |
+| `char`    | `u16`     | Note the usage of `u16`, NOT `u32`. |
```
Depending which language you come from, you'd expect a different type here: `char` would be `u8` coming from C, `u16` from Java, or `u32` from Rust. AIDL picked the Java option.

randomPoison commented on src/android/aidl/birthday_service/src/lib.rs:
```diff
@@ -26,4 +30,64 @@ impl IBirthdayService for BirthdayService {
+        use binder::binder_impl::Proxy;
+
+        // Convert the `IBinder` to a concrete interface.
+        let provider = BpBirthdayInfoProvider::from_binder(provider.clone())?;
```
I'm not sure i understand your suggestion. `IBirthdayInfoProvider` is a trait, so there's no way to impl `TryFrom` for it. I tried doing `BpBirthdayInfoProvider::try_from` in case that's what you meant, but the compiler says there's no `TryFrom` impl for that type.

ahomescu replied:
I saw `FromIBinder::try_from(self)` used at https://android.googlesource.com/platform//frameworks/native/+/refs/heads/main/libs/binder/rust/src/proxy.rs#108, but maybe that should have been `I::try_from`.

Alternatively, you can use the `into_interface` method defined there. See https://android.googlesource.com/platform/system/tools/aidl/+/refs/heads/main/tests/rust/test_client.rs#1054 for an example.

randomPoison replied:
`into_inteface` worked! So that unnecessary import is gone.

ahomescu reviewed (commented):
LGTM, looks like I'm not allowed to approve the PR.

thedataking reviewed (approved):
Approving on behalf of Andrei 🚀

mgeisler commented:
Thanks all for the much-needed updates here! :smile: 

Two ideas:

- Can we link this from our internal Binder documentation? I feel this section has now become a quite nice reference.
- Can we figure out a way to get this tested in CI? Perhaps you could prepare a Docker image and use this somehow?
  - This is the first hit: https://github.com/kylemanna/docker-aosp, but it looks like it hasn't been touched in ~7 years
  - Another hit: https://android.googlesource.com/platform/build/+/master/tools/docker, but this has also not been touched in ~4 years
